File: packages/ios/Sources/CivUI/CivSelect.swift
// CivUI — CivSelect for SwiftUI
// Accessible dropdown select following government design system patterns.
// Renders: label → hint → error → picker (Section 508 compliant)

import SwiftUI

/// A single option for CivSelect (parallels web `SelectOption` interface).
public struct CivSelectOption: Identifiable, Hashable {
    public let id: String
    public let value: String
    public let label: String
    public let isDisabled: Bool

    public init(value: String, label: String, isDisabled: Bool = false) {
        self.id = value
        self.value = value
        self.label = label
        self.isDisabled = isDisabled
    }
}

/// Accessible dropdown select for government applications.
///
/// Implements the CivUI form field pattern with:
/// - Visible label (required for Section 508)
/// - Optional hint text
/// - Error message with VoiceOver announcement
/// - Empty/placeholder option
/// - Dark mode adaptive colors
///
/// Usage:
/// ```swift
/// CivSelect(
///     label: "State",
///     value: $state,
///     options: [
///         CivSelectOption(value: "CA", label: "California"),
///         CivSelectOption(value: "NY", label: "New York"),
///     ]
/// )
/// ```
public struct CivSelect: View {
    // MARK: - Properties

    /// Visible label text (required — never use placeholder as sole label).
    public let label: String

    /// Bound selected value.
    @Binding public var value: String

    /// Available options.
    public var options: [CivSelectOption]

    /// Help text shown below the label.
    public var hint: String?

    /// Error message. When set, renders with VoiceOver announcement.
    public var error: String?

    /// Whether the field is required.
    public var isRequired: Bool

    /// Whether the field is disabled.
    public var isDisabled: Bool

    /// Whether the field is read-only (visible but non-interactive, no opacity change).
    public var isReadonly: Bool

    /// Label for the empty/default option (parallels web `empty-label`).
    public var emptyLabel: String

    /// Called on value change (parallels `civ-input` and `civ-change` events).
    public var onChange: ((String) -> Void)?

    /// Called for analytics tracking (parallels `civ-analytics` event).
    public var onAnalytics: ((String, [String: Any]?) -> Void)?

    /// Optional form state for centralized validation.
    public var formState: CivFormState?

    /// Field name for form state registration.
    public var formName: String?

    /// Custom required message for form validation.
    public var requiredMessage: String?

    /// Custom validation function. Returns error string or nil.
    public var formValidate: (() -> String?)?

    /// Whether the field contains PII (excluded from getFormData).
    public var isPii: Bool

    /// Width variant for the select field (e.g., "full", "auto").
    public var width: String

    /// Density variant — "default" or "sm" (compact). Suppresses chrome
    /// for use in dense surfaces like data-grid cell editors.
    public var spacing: String

    /// Autocomplete hint for the browser/OS.
    public var autocomplete: String

    /// Preset option set name (e.g., "us-states", "countries").
    public var preset: String

    /// Preset variant (e.g., "abbreviated", "full").
    public var presetVariant: String

    // MARK: - Internal State

    @Environment(\.colorScheme) private var colorScheme

    // MARK: - Initializer

    public init(
        label: String,
        value: Binding<String>,
        options: [CivSelectOption],
        hint: String? = nil,
        error: String? = nil,
        isRequired: Bool = false,
        isDisabled: Bool = false,
        isReadonly: Bool = false,
        emptyLabel: String? = nil,
        onChange: ((String) -> Void)? = nil,
        onAnalytics: ((String, [String: Any]?) -> Void)? = nil,
        formState: CivFormState? = nil,
        formName: String? = nil,
        requiredMessage: String? = nil,
        formValidate: (() -> String?)? = nil,
        isPii: Bool = false,
        width: String = "full",
        spacing: String = "default",
        autocomplete: String = "",
        preset: String = "",
        presetVariant: String = ""
    ) {
        self.label = label
        self._value = value
        self.options = options
        self.hint = hint
        self.error = error
        self.isRequired = isRequired
        self.isDisabled = isDisabled
        self.isReadonly = isReadonly
        self.emptyLabel = emptyLabel ?? CivLocale.shared.t("selectEmpty")
        self.onChange = onChange
        self.onAnalytics = onAnalytics
        self.formState = formState
        self.formName = formName
        self.requiredMessage = requiredMessage
        self.formValidate = formValidate
        self.isPii = isPii
        self.width = width
        self.spacing = spacing
        self.autocomplete = autocomplete
        self.preset = preset
        self.presetVariant = presetVariant
    }

    // MARK: - Body

    public var body: some View {
        VStack(alignment: .leading, spacing: CivTokens.Spacing._1) {
            // 1. Label
            labelView

            // 2. Hint
            if let hint, !hint.isEmpty {
                Text(hint)
                    .font(.system(size: CivTokens.Typography.FontSize.sm))
                    .foregroundColor(adaptiveColor(
                        light: CivTokens.Colors.Base.dark,
                        dark: CivTokens.DarkColors.Base.dark
                    ))
                    .accessibilityIdentifier("civ-hint")
            }

            // 3. Error
            if let error, !error.isEmpty {
                Text(error)
                    .font(.system(size: CivTokens.Typography.FontSize.sm,
                                  weight: CivTokens.Typography.FontWeight.bold))
                    .foregroundColor(adaptiveColor(
                        light: CivTokens.Colors.Error.default_,
                        dark: CivTokens.DarkColors.Error.default_
                    ))
                    .accessibilityIdentifier("civ-error")
            }

            // 4. Picker
            pickerField
        }
        .padding(.bottom, CivTokens.Spacing._4)
        .accessibilityElement(children: .contain)
        .onChange(of: error) { newError in
            if let newError, !newError.isEmpty {
                UIAccessibility.post(notification: .announcement, argument: newError)
            }
        }
        .onAppear { registerWithFormState() }
        .onDisappear { unregisterFromFormState() }
    }

    // MARK: - Form State Registration

    private func registerWithFormState() {
        guard let formState, let formName, !formName.isEmpty else { return }
        formState.register(CivFormState.CivFieldRegistration(
            name: formName,
            label: label,
            getValue: { value },
            setValue: { _ in },
            isRequired: isRequired,
            requiredMessage: requiredMessage ?? "",
            validate: formValidate,
            isPii: isPii
        ))
    }

    private func unregisterFromFormState() {
        guard let formState, let formName, !formName.isEmpty else { return }
        formState.unregister(formName)
    }

    // MARK: - Subviews

    private var labelView: some View {
        HStack(spacing: CivTokens.Spacing._0_5) {
            Text(label)
                .font(.system(size: CivTokens.Typography.FontSize.base,
                              weight: CivTokens.Typography.FontWeight.bold))
                .foregroundColor(adaptiveColor(
                    light: CivTokens.Colors.Base.darkest,
                    dark: CivTokens.DarkColors.Base.darkest
                ))

            if isRequired {
                Text("*")
                    .font(.system(size: CivTokens.Typography.FontSize.base,
                                  weight: CivTokens.Typography.FontWeight.bold))
                    .foregroundColor(adaptiveColor(
                        light: CivTokens.Colors.Error.default_,
                        dark: CivTokens.DarkColors.Error.default_
                    ))
                    .accessibilityLabel("required")
            }
        }
    }

    private var pickerField: some View {
        Picker(selection: selectBinding, label: EmptyView()) {
            Text(emptyLabel).tag("")
            ForEach(options) { option in
                Text(option.label)
                    .tag(option.value)
            }
        }
        .pickerStyle(.menu)
        .font(.system(size: CivTokens.Typography.FontSize.base))
        .padding(.horizontal, CivTokens.Spacing._2)
        .padding(.vertical, CivTokens.Spacing._1_5)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(adaptiveColor(
            light: CivTokens.Colors.White.default_,
            dark: CivTokens.DarkColors.White.default_
        ))
        .cornerRadius(CivTokens.Border.Radius.default_)
        .overlay(borderOverlay)
        .disabled(isDisabled || isReadonly)
        .opacity(isDisabled ? 0.5 : 1.0)
        .accessibilityLabel(accessibilityLabelText)
        .accessibilityHint(accessibilityHintText)
    }

    // MARK: - Border

    private var borderOverlay: some View {
        RoundedRectangle(cornerRadius: CivTokens.Border.Radius.default_)
            .stroke(borderColor, lineWidth: error != nil
                ? CivTokens.Border.Width._2
                : CivTokens.Border.Width.default_)
    }

    private var borderColor: Color {
        if error != nil {
            return adaptiveColor(
                light: CivTokens.Colors.Error.default_,
                dark: CivTokens.DarkColors.Error.default_
            )
        }
        return adaptiveColor(
            light: CivTokens.Colors.Base.light,
            dark: CivTokens.DarkColors.Base.light
        )
    }

    // MARK: - Select Binding

    private var selectBinding: Binding<String> {
        Binding(
            get: { value },
            set: { newValue in
                value = newValue
                onChange?(newValue)
                onAnalytics?("change", ["value": newValue])
            }
        )
    }

    // MARK: - Accessibility

    private var accessibilityLabelText: String {
        var parts = [label]
        if isRequired { parts.append("required") }
        return parts.joined(separator: ", ")
    }

    private var accessibilityHintText: String {
        var parts: [String] = []
        if let hint, !hint.isEmpty { parts.append(hint) }
        if let error, !error.isEmpty { parts.append("Error: \(error)") }
        return parts.joined(separator: ". ")
    }

    // MARK: - Color Helper

    private func adaptiveColor(light: Color, dark: Color) -> Color {
        colorScheme == .dark ? dark : light
    }
}

// MARK: - Preview

#if DEBUG
struct CivSelect_Previews: PreviewProvider {
    struct PreviewWrapper: View {
        @State private var state = ""
        @State private var branch = "executive"

        var body: some View {
            ScrollView {
                VStack(alignment: .leading, spacing: 0) {
                    CivSelect(
                        label: "State or territory",
                        value: $state,
                        options: [
                            CivSelectOption(value: "CA", label: "California"),
                            CivSelectOption(value: "DC", label: "District of Columbia"),
                            CivSelectOption(value: "NY", label: "New York"),
                            CivSelectOption(value: "TX", label: "Texas"),
                        ],
                        hint: "Select your state of residence",
                        isRequired: true
                    )

                    CivSelect(
                        label: "Branch of government",
                        value: $branch,
                        options: [
                            CivSelectOption(value: "executive", label: "Executive"),
                            CivSelectOption(value: "legislative", label: "Legislative"),
                            CivSelectOption(value: "judicial", label: "Judicial"),
                        ]
                    )

                    CivSelect(
                        label: "Agency",
                        value: .constant(""),
                        options: [],
                        error: "Please select an agency",
                        isRequired: true
                    )
                }
                .padding()
            }
        }
    }

    static var previews: some View {
        PreviewWrapper()
            .previewDisplayName("Light")
        PreviewWrapper()
            .preferredColorScheme(.dark)
            .previewDisplayName("Dark")
    }
}
#endif

File: packages/android/src/main/kotlin/gov/civui/components/CivSelect.kt
// CivUI — CivSelect for Jetpack Compose
// Accessible dropdown select following government design system patterns.
// Renders: label -> hint -> error -> dropdown (Section 508 compliant)

package gov.civui.components

import androidx.compose.animation.animateColorAsState
import androidx.compose.foundation.border
import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.ExposedDropdownMenuBox
import androidx.compose.material3.ExposedDropdownMenuDefaults
import androidx.compose.material3.MenuAnchorType
import androidx.compose.material3.Text
import androidx.compose.material3.TextField
import androidx.compose.material3.TextFieldDefaults
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.error
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import gov.civui.i18n.CivLocale
import gov.civui.tokens.CivTokens

/**
 * Option item for CivSelect (parallels web SelectOption).
 */
data class CivSelectOption(
    val value: String,
    val label: String,
    val disabled: Boolean = false,
)

/**
 * Accessible dropdown select for government applications.
 *
 * Uses Material 3 ExposedDropdownMenuBox for native behavior,
 * with CivUI styling and accessibility patterns.
 *
 * Usage:
 * ```kotlin
 * var state by remember { mutableStateOf("") }
 * CivSelect(
 *     label = "State or territory",
 *     value = state,
 *     onValueChange = { state = it },
 *     options = listOf(
 *         CivSelectOption("DC", "District of Columbia"),
 *         CivSelectOption("MD", "Maryland"),
 *         CivSelectOption("VA", "Virginia"),
 *     ),
 * )
 * ```
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun CivSelect(
    label: String,
    value: String,
    onValueChange: (String) -> Unit,
    options: List<CivSelectOption>,
    modifier: Modifier = Modifier,
    hint: String = "",
    error: String = "",
    required: Boolean = false,
    disabled: Boolean = false,
    emptyLabel: String = CivLocale.t("selectEmpty"),
    name: String = "",
    formState: CivFormState? = null,
    onChange: ((String) -> Unit)? = null,
    onAnalytics: ((event: String, data: Map<String, Any>?) -> Unit)? = null,
) {
    val isDark = isSystemInDarkTheme()

    // Form state registration
    var formError by remember { mutableStateOf("") }
    val effectiveError = error.ifEmpty { formError }

    if (formState != null && name.isNotEmpty()) {
        androidx.compose.runtime.DisposableEffect(name) {
            formState.register(CivFormState.CivFieldRegistration(
                name = name,
                getValue = { value },
                setValue = { onValueChange(it) },
                isRequired = required,
                getError = { formError },
                setError = { formError = it },
            ))
            onDispose { formState.unregister(name) }
        }
    }

    val labelColor = if (isDark) CivTokens.DarkColors.Base.darkest else CivTokens.Colors.Base.darkest
    val hintColor = if (isDark) CivTokens.DarkColors.Base.dark else CivTokens.Colors.Base.dark
    val errorColor = if (isDark) CivTokens.DarkColors.Error.default_ else CivTokens.Colors.Error.default_
    val borderColor by animateColorAsState(
        targetValue = when {
            effectiveError.isNotEmpty() -> errorColor
            else -> if (isDark) CivTokens.DarkColors.Base.light else CivTokens.Colors.Base.light
        },
        label = "borderColor",
    )
    val backgroundColor = if (isDark) CivTokens.DarkColors.White.default_ else CivTokens.Colors.White.default_

    var expanded by remember { mutableStateOf(false) }

    // Find the label for the current value
    val selectedLabel = options.find { it.value == value }?.label ?: ""

    Column(
        modifier = modifier.padding(bottom = CivTokens.Spacing._4),
    ) {
        // 1. Label
        CivLabel(
            label = label,
            required = required,
            labelColor = labelColor,
            errorColor = errorColor,
        )

        // 2. Hint
        CivHint(
            text = hint.ifEmpty { null },
            color = hintColor,
        )

        // 3. Error
        CivError(
            text = effectiveError.ifEmpty { null },
            color = errorColor,
        )

        // 4. Dropdown
        ExposedDropdownMenuBox(
            expanded = expanded,
            onExpandedChange = { if (!disabled) expanded = it },
        ) {
            TextField(
                value = selectedLabel.ifEmpty { emptyLabel },
                onValueChange = {},
                readOnly = true,
                modifier = Modifier
                    .menuAnchor(MenuAnchorType.PrimaryNotEditable)
                    .fillMaxWidth()
                    .border(
                        width = if (effectiveError.isNotEmpty()) CivTokens.Border.Width._2 else CivTokens.Border.Width.default_,
                        color = borderColor,
                        shape = RoundedCornerShape(CivTokens.Border.Radius.default_),
                    )
                    .alpha(if (disabled) 0.5f else 1f)
                    .semantics {
                        contentDescription = buildString {
                            append(label)
                            if (required) append(", required")
                            if (selectedLabel.isNotEmpty()) append(", selected: $selectedLabel")
                            if (hint.isNotEmpty()) append(". $hint")
                            if (effectiveError.isNotEmpty()) append(". Error: $effectiveError")
                        }
                        if (effectiveError.isNotEmpty()) {
                            error(effectiveError)
                        }
                    },
                enabled = !disabled,
                textStyle = TextStyle(
                    fontSize = CivTokens.Typography.FontSize.base,
                    color = if (selectedLabel.isEmpty()) hintColor else labelColor,
                ),
                trailingIcon = { ExposedDropdownMenuDefaults.TrailingIcon(expanded = expanded) },
                shape = RoundedCornerShape(CivTokens.Border.Radius.default_),
                colors = TextFieldDefaults.colors(
                    focusedContainerColor = backgroundColor,
                    unfocusedContainerColor = backgroundColor,
                    disabledContainerColor = backgroundColor,
                    focusedIndicatorColor = Color.Transparent,
                    unfocusedIndicatorColor = Color.Transparent,
                    disabledIndicatorColor = Color.Transparent,
                ),
            )

            ExposedDropdownMenu(
                expanded = expanded,
                onDismissRequest = { expanded = false },
            ) {
                // Empty option
                DropdownMenuItem(
                    text = {
                        Text(
                            text = emptyLabel,
                            style = TextStyle(
                                fontSize = CivTokens.Typography.FontSize.base,
                                color = hintColor,
                            ),
                        )
                    },
                    onClick = {
                        onValueChange("")
                        onChange?.invoke("")
                        onAnalytics?.invoke("change", mapOf("field" to label, "value" to ""))
                        expanded = false
                    },
                )

                // Options
                options.forEach { option ->
                    DropdownMenuItem(
                        text = {
                            Text(
                                text = option.label,
                                style = TextStyle(
                                    fontSize = CivTokens.Typography.FontSize.base,
                                    color = labelColor,
                                ),
                            )
                        },
                        onClick = {
                            if (!option.disabled) {
                                onValueChange(option.value)
                                onChange?.invoke(option.value)
                                onAnalytics?.invoke("change", mapOf("field" to label, "value" to option.value))
                                expanded = false
                            }
                        },
                        enabled = !option.disabled,
                    )
                }
            }
        }
    }
}

// MARK: - Preview

@Preview(showBackground = true, name = "CivSelect")
@Composable
private fun CivSelectPreview() {
    Column(modifier = Modifier.padding(16.dp)) {
        var state by remember { mutableStateOf("") }
        CivSelect(
            label = "State or territory",
            value = state,
            onValueChange = { state = it },
            options = listOf(
                CivSelectOption("DC", "District of Columbia"),
                CivSelectOption("MD", "Maryland"),
                CivSelectOption("VA", "Virginia"),
            ),
            hint = "Select your state of residence",
            required = true,
        )

        CivSelect(
            label = "Disabled select",
            value = "",
            onValueChange = {},
            options = emptyList(),
            disabled = true,
        )
    }
}
